### ISSUE TITLE
Mobility stores `nil` instead of a JSONB object if the attribute is set via #write_attribute (JSONB backend).

### ISSUE THREAD

**Issue by smaximov** [shioyama/mobility#99]
Not really sure that it is a Mobility issue, it's more likely a compatibility bug with [strip_attributes][]. If an attribute was previously set via `#write_attribute` (or the `#[]=` alias), when AR persists changes, this attribute is stored as nil (instead of a JSONB object `{ locale => attribute_value }`).

## Context

I use `mobility` and [strip_attributes][] for my project. Recently I started to notice that sometimes when the user updates a record with columns backed by the Mobility's JSONB backend, some of these columns' values disappear (corresponding Postgres JSONB field is set to NULL, which results in losing all available translations for that field). After some debugging I found out that it happens when an attribute's value contains leading or trailing whitespace and the model's attributes are stripped via `strip_attributes`. 

`strip_attributes` basically registers a `before_validation` hook that updates an attribute via `#[]=` if the stripped value differs from the original value, hence the reproducible example:

``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required.'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'

  gem 'activerecord', '5.1.0'
  gem 'mobility', '0.2.2'
  gem 'pg', '0.18.4'
  gem 'rspec', '3.6.0', require: 'rspec/autorun'
  gem 'pry-byebug'
end

DB = ENV.fetch('MOBILITY_TEST_DB', 'mobility-test').freeze

ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: DB)
ActiveRecord::Base.logger = Logger.new($stdout)

I18n.locale = I18n.default_locale = :en
I18n.available_locales = %i[en ru]

Mobility.configure do |config|
  config.default_backend = :jsonb
  config.accessor_method = :translates
end

ActiveRecord::Schema.define do
  create_table :articles, force: true do |t|
    t.jsonb :title, default: {}
    t.jsonb :title_strip, default: {}
  end
end

class Article < ActiveRecord::Base
  extend Mobility

  translates :title, :title_strip

  STRIP_ATTRIBUTES = %i[title_strip].freeze

  before_validation do |record|
    record.attributes.slice(*STRIP_ATTRIBUTES.map(&:to_s)).each do |attr, value|
      next unless value.respond_to?(:strip)
      stripped_value = value.strip
      record[attr] = stripped_value if value != stripped_value
    end
  end
end

RSpec.describe Article do
  let(:params) { Hash[title: title, title_strip: title] }

  subject(:article) { Article.create!(params) }

  describe '#create!' do
    context "when title doesn't have leading or trailing space" do
      let(:title) { 'title without leading or trailing space' }

      it { expect(article.title).to eq(title) }
      it { expect(article.title_strip).to eq(title) }
    end

    context 'when title has leading space' do
      let(:title) { ' title with leading space' }

      it { expect(article.title).to eq(title) }
      it { expect(article.title_strip).to eq(title.strip) }
    end
  end
end
```

## Expected Behavior
The tests should pass.

## Actual Behavior
The last test fails.

## Possible Fix

How I can mitigate the problem:

* move input pre-processing out of the model;
* blacklist Mobility-backed attributes for `strip_attributes` via the `:except` options (this is what I've done so far).

How it can be fixed:

* make `strip_attributes` use attribute accessor methods for writing to a model (use `record.public_send("#{attr}=", value)` instead of `record[attr] = value`;
* ???

[strip_attributes]: https://github.com/rmm5t/strip_attributes

**Comment by shioyama** [comment]
This overlaps with #97 (compatibility with strip_attributes) and #91 (different column for jsonb backend). As I understand it, you're actually encountering both issues combined. The solution is:

1) to remove the code in Mobility which [adds translated attributes](https://github.com/shioyama/mobility/blob/e803847f967117a34ef8f31491af555836d97c6d/lib/mobility/active_record/attribute_methods.rb#L29-L33) to the `attributes` hash. This will prevent strip_attributes from applying attribute stripping to these attributes, which is one source of problems.

In the case of jsonb, there is another problem discussed in #91, which is that Mobility uses the column as the attribute name. This also contributes to a compatibility issue with strip_attributes. The proposal in #91 is to name the column something different from the attribute name, which I'm now considering. This would be a breaking change, however, so wouldn't be released until the next minor version (0.3.0).

I need a bit more time to look into these issues. The first one is very easy (just removing the code), and the second one is not hard to implement but requires some careful thought about what default to use.

I'd like to close this since I believe that the other issues will cover your case here. Please have a look at them and see if you agree.

**Comment by smaximov** [comment]
> Please have a look at them and see if you agree.

Yes, I totally agree. Thanks for the prompt reply!
